<a href="https://colab.research.google.com/github/achalbajpai/learning-about-llms/blob/main/LLMsFromScratch.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **BUILDING LLM FROM SCRATCH : UPDATED TILL JANUARY 31**




## Step 1: Creating Tokens

<div class="alert alert-block alert-success">

The print command prints the total number of characters followed by the first 100
characters of this file for illustration purposes. </div>

In [ ]:
# Open the file "cristiano-ronaldo.txt" in read mode with UTF-8 encoding
# The `with` statement ensures the file is properly closed after reading
with open("cr7.txt", "r", encoding="utf-8") as f:
    # Read the entire content of the file into the variable `raw_text`
    raw_text = f.read()

# Print the total number of characters in the file
# `len(raw_text)` returns the length of the string `raw_text`, which is the number of characters
print("Total number of characters:", len(raw_text))

# Print the first 99 characters of the file
# `raw_text[:99]` slices the string `raw_text` to get the first 99 characters
t

Total number of characters: 4348
Cristiano Ronaldo dos Santos Aveiro, born on 5 February 1985, is a Portuguese professional football


# Tokenizer


<div class="alert alert-block alert-success">

Our goal is to tokenize this 4348-character short story into individual words and special
characters that we can then turn into embeddings for LLM training  </div>

<div class="alert alert-block alert-warning">

Note that it's common to process millions of articles and hundreds of thousands of
books -- many gigabytes of text -- when working with LLMs. However, for educational
purposes, it's sufficient to work with smaller text samples like a single book to
illustrate the main ideas behind the text processing steps and to make it possible to
run it in reasonable time on consumer hardware. </div>

<div class="alert alert-block alert-success">

How can we best split this text to obtain a list of tokens? For this, we go on a small
excursion and use Python's regular expression library re for illustration purposes. (Note
that you don't have to learn or memorize any regular expression syntax since we will
transition to a pre-built tokenizer later in this chapter.) </div>

In [ ]:
# Import the `re` module, which provides support for regular expressions in Python
import re

# Define a sample text string to be processed
text = "Hello, world. This, is a test."

# Use the `re.split()` function to split the text into tokens and separators
# r makes the string as a RAW string aka paranthesis get ignored in it as they are regex expression
# The regular expression `([,.:;?_!"()\']|--|\s)` matches:
# - Punctuation marks: , . : ; ? _ ! " ( ) '
# - Double hyphens: --
# - Whitespace: \s
# The parentheses `()` around the regex ensure that the separators are included in the result
result = re.split(r'([,.:;?_!"()\']|--|\s)', text)

# Print the initial result of the split
print(result)


['Hello', ',', '', ' ', 'world', '.', '', ' ', 'This', ',', '', ' ', 'is', ' ', 'a', ' ', 'test', '.', '']


<div class="alert alert-block alert-info">
The result is a list of individual words, whitespaces, and punctuation characters:
</div>


<div class="alert alert-block alert-warning">

Let's modify the regular expression splits on whitespaces (\s) and commas, and periods
([,.]):</div>

In [ ]:
# Clean the result list:
# 1. Use a list comprehension to iterate over each item in `result`
# 2. Apply `item.strip()` to remove leading and trailing whitespace from each item
# 3. Filter out any empty strings using `if item.strip()`
result = [item.strip() for item in result if item.strip()]

# Print the cleaned result
print(result)


['Hello', ',', 'world', '.', 'This', ',', 'is', 'a', 'test', '.']


<div class="alert alert-block alert-success">

REMOVING WHITESPACES OR NOT


When developing a simple tokenizer, whether we should encode whitespaces as
separate characters or just remove them depends on our application and its
requirements. Removing whitespaces reduces the memory and computing
requirements. However, keeping whitespaces can be useful if we train models that
are sensitive to the exact structure of the text (for example, Python code, which is
sensitive to indentation and spacing). Here, we remove whitespaces for simplicity
and brevity of the tokenized outputs. Later, we will switch to a tokenization scheme
that includes whitespaces.

</div>

<div class="alert alert-block alert-warning">

The tokenization scheme we devised above works well on the simple sample text. Let's
modify it a bit further so that it can also handle other types of punctuation, such as
question marks, quotation marks, and the double-dashes we have seen earlier in the first
100 characters of Edith Wharton's short story, along with additional special characters: </div>

In [ ]:
text = "Hello, world. Is this-- a test?"
result = re.split(r'([,.:;?_!"()\']|--|\s)', text)
result = [item.strip() for item in result if item.strip()]
print(result)

['Hello', ',', 'world', '.', 'Is', 'this', '--', 'a', 'test', '?']


In [ ]:
# Alternative approach (redundant in this case, as the previous step already handles this):
# Filter out any empty strings from the list
# This step is unnecessary here because the previous list comprehension already removes empty strings
result = [item for item in result if item.strip()]

# Print the final result
print(result)

['Hello', ',', 'world', '.', 'Is', 'this', '--', 'a', 'test', '?']


In [ ]:
# Use the `re.split()` function to split the `raw_text` string into tokens and separators
# The regular expression `([,.:;?_!"()\']|--|\s)` matches:
# - Punctuation marks: , . : ; ? _ ! " ( ) '
# - Double hyphens: --
# - Whitespace: \s
# The parentheses `()` around the regex ensure that the separators are included in the result
preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', raw_text)

# Clean the `preprocessed` list:
# 1. Use a list comprehension to iterate over each item in `preprocessed`
# 2. Apply `item.strip()` to remove leading and trailing whitespace from each item
# 3. Filter out any empty strings using `if item.strip()`
preprocessed = [item.strip() for item in preprocessed if item.strip()]

# Print the first 30 items of the cleaned `preprocessed` list
# This is useful for inspecting the initial part of the processed data
print(preprocessed[:30])

# Print the total number of items in the cleaned `preprocessed` list
# This gives an idea of the size of the processed data
print(len(preprocessed))

['Cristiano', 'Ronaldo', 'dos', 'Santos', 'Aveiro', ',', 'born', 'on', '5', 'February', '1985', ',', 'is', 'a', 'Portuguese', 'professional', 'footballer', '.', 'He', 'plays', 'as', 'a', 'forward', 'for', 'and', 'captains', 'both', 'the', 'Saudi', 'Pro']
882


# **Step 2 - Token to Token ID**


<div class="alert alert-block alert-warning">

In the previous section, we tokenized CR7 short story and assigned it to a
Python variable called preprocessed. Let's now create a list of all unique tokens and sort
them alphabetically to determine the vocabulary size:</div>

In [ ]:
# Create a sorted list of unique words (tokens) from the `preprocessed` list
# 1. `set(preprocessed)` creates a set of unique items from `preprocessed`.
# 2. `sorted()` sorts the unique items alphabetically.
all_words = sorted(set(preprocessed))

# Calculate the size of the vocabulary (number of unique words)
# `len(all_words)` returns the number of items in the `all_words` list
vocab_size = len(all_words)

# Print the size of the vocabulary
# This gives the total number of unique words in the preprocessed data
print(vocab_size)

303


<div class="alert alert-block alert-success">

After determining that the vocabulary size is 303 via the above code, we create the
vocabulary and print its first 51 entries for illustration purposes:

</div>

In [ ]:
# Create a vocabulary dictionary that maps each unique token to a unique integer
# 1. `enumerate(all_words)` generates pairs of (index, token) for each token in `all_words`.
#    - `index` is the position of the token in the sorted list (starting from 0).
#    - `token` is the word itself.
# 2. A dictionary comprehension is used to create the `vocab` dictionary:
#    - The key is the `token` (word).
#    - The value is the `integer` (index).
vocab = {token: integer for integer, token in enumerate(all_words)}

In [ ]:
# Iterate over the items in the `vocab` dictionary using `enumerate`
# `enumerate(vocab.items())` generates pairs of (index, (token, integer)) for each item in `vocab`
# - `index` is the position of the item in the iteration (starting from 0).
# - `item` is a tuple of (token, integer), where:
#   - `token` is the word (key in the dictionary).
#   - `integer` is the corresponding ID (value in the dictionary).
for i, item in enumerate(vocab.items()):
    # Print the current item (tuple of (token, integer))
    print(item)

    # Stop the loop after printing 50 items
    # This is useful for inspecting the first few items in a large dictionary
    if i >= 50:
        break

("'", 0)
('(', 1)
(')', 2)
(',', 3)
('.', 4)
('1', 5)
('100', 6)
('135', 7)
('14', 8)
('140', 9)
('18', 10)
('183', 11)
('1985', 12)
('200', 13)
('2003', 14)
('2004', 15)
('2008', 16)
('2009', 17)
('2013', 18)
('2014', 19)
('2015', 20)
('2016', 21)
('2017', 22)
('2018', 23)
('2019', 24)
('2020', 25)
('2021', 26)
('2022', 27)
('2023', 28)
('2024', 29)
('217', 30)
('23', 31)
('30', 32)
('33', 33)
('42', 34)
('5', 35)
('8', 36)
('900', 37)
('A', 38)
('Additionally', 39)
('Al', 40)
('At', 41)
('Aveiro', 42)
('Awards', 43)
('Bale', 44)
('Ballon', 45)
('Ballons', 46)
('Benzema', 47)
('Boot', 48)
('CP', 49)
('Champions', 50)


<div class="alert alert-block alert-info">
As we can see, based on the output above, the dictionary contains individual tokens
associated with unique integer labels.
</div>

<div class="alert alert-block alert-success">

Later in this book, when we want to convert the outputs of an LLM from numbers back into
text, we also need a way to turn token IDs into text.

For this, we can create an inverse
version of the vocabulary that maps token IDs back to corresponding text tokens.

</div>

<div class="alert alert-block alert-success">

Let's implement a complete tokenizer class in Python.

The class will have an encode method that splits
text into tokens and carries out the string-to-integer mapping to produce token IDs via the
vocabulary.

In addition, we implement a decode method that carries out the reverse
integer-to-string mapping to convert the token IDs back into text.

</div>

<div class="alert alert-block alert-info">
    
Step 1: Store the vocabulary as a class attribute for access in the encode and decode methods
    
Step 2: Create an inverse vocabulary that maps token IDs back to the original text tokens

Step 3: Process input text into token IDs

Step 4: Convert token IDs back into text

Step 5: Replace spaces before the specified punctuation

</div>



In [ ]:
class SimpleTokenizerV1:
    def __init__(self, vocab):
        """
        Initialize the tokenizer with a vocabulary mapping.

        Args:
            vocab (dict): A dictionary mapping tokens (strings) to unique integers (IDs).
        """
        # Map tokens to IDs (e.g., {"Hello": 1, "world": 2})
        self.str_to_int = vocab

        # Create the inverse mapping: IDs to tokens (e.g., {1: "Hello", 2: "world"})
        self.int_to_str = {i: s for s, i in vocab.items()}

    def encode(self, text):
        """
        Convert a text string into a list of token IDs.

        Args:
            text (str): The input text to encode.

        Returns:
            list: A list of token IDs corresponding to the input text.
        """
        import re

        # Step 1: Preprocess the text
        # Split the text into tokens and separators (like punctuation and spaces)
        # using a regular expression.
        preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', text)

        # Step 2: Clean the preprocessed list
        # Remove leading/trailing spaces from each item and filter out empty strings.
        preprocessed = [
            item.strip() for item in preprocessed if item.strip()
        ]

        # Step 3: Convert tokens to IDs
        # Use the str_to_int mapping to convert each token to its corresponding ID.
        ids = [self.str_to_int[s] for s in preprocessed]

        # Return the list of token IDs
        return ids

    def decode(self, ids):
        """
        Convert a list of token IDs back into a text string.

        Args:
            ids (list): A list of token IDs to decode.

        Returns:
            str: The decoded text string.
        """
        # Step 1: Convert IDs back to tokens
        # Use the int_to_str mapping to convert each ID to its corresponding token.
        tokens = [self.int_to_str[i] for i in ids]

        # Step 2: Join tokens into a single string with spaces in between
        text = " ".join(tokens)

        # Step 3: Postprocess the text
        # Fix spacing issues around punctuation marks (e.g., "Hello , world !" -> "Hello, world!")
        text = re.sub(r'\s+([,.?!"()\'])', r'\1', text)

        # Return the decoded text
        return text

<div class="alert alert-block alert-success">

Let's instantiate a new tokenizer object from the SimpleTokenizerV1 class and tokenize a
passage from CR7 short story to try it out in practice:
</div>

In [ ]:
tokenizer = SimpleTokenizerV1(vocab)

# Input text
text = """Ronaldo has won five Ballon d'Or awards, a record three UEFA Men's Player of the Year Awards, and four European Golden Shoes. He was named the world's best player by FIFA five times, the most by a European player. He has won 33 trophies in his career, including seven league titles, five UEFA Champions Leagues, the UEFA European Championship, and the UEFA Nations League."""

# Encode the text
ids = tokenizer.encode(text)
print(ids)

[100, 191, 296, 177, 45, 163, 0, 92, 144, 3, 123, 249, 273, 114, 88, 0, 253, 93, 230, 268, 121, 43, 3, 134, 184, 63, 72, 104, 4, 73, 289, 227, 268, 297, 0, 253, 149, 240, 155, 65, 177, 276, 3, 268, 225, 155, 123, 63, 240, 4, 73, 191, 296, 33, 286, 200, 197, 159, 3, 203, 259, 211, 278, 3, 177, 114, 50, 84, 3, 268, 114, 63, 51, 3, 134, 268, 114, 91, 83, 4]


In [ ]:
tokenizer.decode(ids)


"Ronaldo has won five Ballon d' Or awards, a record three UEFA Men' s Player of the Year Awards, and four European Golden Shoes. He was named the world' s best player by FIFA five times, the most by a European player. He has won 33 trophies in his career, including seven league titles, five UEFA Champions Leagues, the UEFA European Championship, and the UEFA Nations League."

<div class="alert alert-block alert-info">
    
Based on the output above, we can see that the decode method successfully converted the
token IDs back into the original text.
</div>

<div class="alert alert-block alert-success">

So far, so good. We implemented a tokenizer capable of tokenizing and de-tokenizing
text based on a snippet from the training set.

Let's now apply it to a new text sample that
is not contained in the training set:
</div>

In [ ]:
# text = "Hello, do you like tea?"
# print(tokenizer.encode(text))

<div class="alert alert-block alert-info">
    
The problem is that the word "Hello" was not used in the The CR7 short story.

Hence, it
is not contained in the vocabulary.

This highlights the need to consider large and diverse
training sets to extend the vocabulary when working on LLMs.

</div>

### ADDING SPECIAL CONTEXT TOKENS

In the previous section, we implemented a simple tokenizer and applied it to a passage
from the training set.

In this section, we will modify this tokenizer to handle unknown
words.


In particular, we will modify the vocabulary and tokenizer we implemented in the
previous section, SimpleTokenizerV2, to support two new tokens, <|unk|> and
<|endoftext|>

<div class="alert alert-block alert-warning">

We can modify the tokenizer to use an <|unk|> token if it
encounters a word that is not part of the vocabulary.

Furthermore, we add a token between
unrelated texts.

For example, when training GPT-like LLMs on multiple independent
documents or books, it is common to insert a token before each document or book that
follows a previous text source

</div>



<div class="alert alert-block alert-success">

Let's now modify the vocabulary to include these two special tokens, <unk> and
<|endoftext|>, by adding these to the list of all unique words that we created in the
previous section:
</div>

In [ ]:
# Create a sorted list of unique tokens from the preprocessed list
# 1. `set(preprocessed)` creates a set of unique tokens from the `preprocessed` list.
# 2. `list(set(preprocessed))` converts the set back into a list.
# 3. `sorted()` sorts the list of unique tokens alphabetically.
all_tokens = sorted(list(set(preprocessed)))

# Add special tokens to the list of all tokens
# 1. `<|endoftext|>` is a special token used to mark the end of a text sequence.
# 2. `<|unk|>` is a special token used to represent unknown or out-of-vocabulary tokens.
all_tokens.extend(["<|endoftext|>", "<|unk|>"])

# Create a vocabulary dictionary that maps each token to a unique integer
# 1. `enumerate(all_tokens)` generates pairs of (index, token) for each token in `all_tokens`.
#    - `index` is the position of the token in the sorted list (starting from 0).
#    - `token` is the word or special token itself.
# 2. A dictionary comprehension is used to create the `vocab` dictionary:
#    - The key is the `token` (word or special token).
#    - The value is the `integer` (index).
vocab = {token: integer for integer, token in enumerate(all_tokens)}

In [ ]:
len(vocab.items())


305

<div class="alert alert-block alert-info">
    
Based on the output of the print statement above, the new vocabulary size is 305 (the
vocabulary size in the previous section was 303).

</div>



<div class="alert alert-block alert-success">

As an additional quick check, let's print the last 5 entries of the updated vocabulary:
</div>

In [ ]:
# Iterate over the last 5 items in the `vocab` dictionary
# 1. `vocab.items()` returns a view of the dictionary's (token, integer) pairs.
# 2. `list(vocab.items())` converts the view into a list of (token, integer) tuples.
# 3. `[-5:]` slices the list to get the last 5 items.
# 4. `enumerate()` adds an index to each item, generating pairs of (index, (token, integer)).
for i, item in enumerate(list(vocab.items())[-5:]):
    # Print the current item (tuple of (token, integer))
    print(item)

('£88', 300)
('€100', 301)
('€94', 302)
('<|endoftext|>', 303)
('<|unk|>', 304)


<div class="alert alert-block alert-success">

A simple text tokenizer that handles unknown words</div>



<div class="alert alert-block alert-info">
    
Step 1: Replace unknown words by <|unk|> tokens
    
Step 2: Replace spaces before the specified punctuations

</div>


In [ ]:
class SimpleTokenizerV2:
    def __init__(self, vocab):
        """
        Initialize the tokenizer with a vocabulary mapping.

        Args:
            vocab (dict): A dictionary mapping tokens (strings) to unique integers (IDs).
        """
        # Map tokens to IDs (e.g., {"Hello": 1, "world": 2})
        self.str_to_int = vocab

        # Create the inverse mapping: IDs to tokens (e.g., {1: "Hello", 2: "world"})
        self.int_to_str = {i: s for s, i in vocab.items()}

    def encode(self, text):
        """
        Convert a text string into a list of token IDs.

        Args:
            text (str): The input text to encode.

        Returns:
            list: A list of token IDs corresponding to the input text.
        """
        import re

        # Step 1: Preprocess the text
        # Split the text into tokens and separators (like punctuation and spaces)
        # using a regular expression.
        preprocessed = re.split(r'([,.:;?_!"()\']|--|\s)', text)

        # Step 2: Clean the preprocessed list
        # Remove leading/trailing spaces from each item and filter out empty strings.
        preprocessed = [item.strip() for item in preprocessed if item.strip()]

        # Step 3: Handle unknown tokens
        # Replace any token not in the vocabulary with the special "<|unk|>" token.
        preprocessed = [
            item if item in self.str_to_int
            else "<|unk|>" for item in preprocessed
        ]

        # Step 4: Convert tokens to IDs
        # Use the str_to_int mapping to convert each token to its corresponding ID.
        ids = [self.str_to_int[s] for s in preprocessed]

        # Return the list of token IDs
        return ids

    def decode(self, ids):
        """
        Convert a list of token IDs back into a text string.

        Args:
            ids (list): A list of token IDs to decode.

        Returns:
            str: The decoded text string.
        """
        # Step 1: Convert IDs back to tokens
        # Use the int_to_str mapping to convert each ID to its corresponding token.
        tokens = [self.int_to_str[i] for i in ids]

        # Step 2: Join tokens into a single string with spaces in between
        text = " ".join(tokens)

        # Step 3: Postprocess the text
        # Fix spacing issues around punctuation marks (e.g., "Hello , world !" -> "Hello, world!")
        text = re.sub(r'\s+([,.:;?!"()\'])', r'\1', text)

        # Return the decoded text
        return text

In [ ]:
# Create a tokenizer instance using the vocabulary
tokenizer = SimpleTokenizerV2(vocab)

# Define two example text strings
text1 = "Hello, do you like tea?"
text2 = "In the sunlit terraces of the palace."

# Combine the two text strings into a single string, separated by the special token "<|endoftext|>"
# The `join()` method inserts "<|endoftext|>" between `text1` and `text2`.
text = " <|endoftext|> ".join((text1, text2))

# Print the combined text
print(text)

Hello, do you like tea? <|endoftext|> In the sunlit terraces of the palace.


In [ ]:
tokenizer.encode(text)


[304, 3, 304, 304, 304, 304, 304, 303, 74, 268, 304, 304, 230, 268, 304, 4]

In [ ]:
tokenizer.decode(tokenizer.encode(text))

'<|unk|>, <|unk|> <|unk|> <|unk|> <|unk|> <|unk|> <|endoftext|> In the <|unk|> <|unk|> of the <|unk|>.'


<div class="alert alert-block alert-info">
    
Based on comparing the de-tokenized text above with the original input text, we know that
the training dataset, CR7 , did not contain the words"

</div>


<div class="alert alert-block alert-warning">

So far, we have discussed tokenization as an essential step in processing text as input to
LLMs. Depending on the LLM, some researchers also consider additional special tokens such
as the following:

[BOS] (beginning of sequence): This token marks the start of a text. It
signifies to the LLM where a piece of content begins.

[EOS] (end of sequence): This token is positioned at the end of a text,
and is especially useful when concatenating multiple unrelated texts,
similar to <|endoftext|>. For instance, when combining two different
Wikipedia articles or books, the [EOS] token indicates where one article
ends and the next one begins.

[PAD] (padding): When training LLMs with batch sizes larger than one,
the batch might contain texts of varying lengths. To ensure all texts have
the same length, the shorter texts are extended or "padded" using the
[PAD] token, up to the length of the longest text in the batch.

</div>


<div class="alert alert-block alert-warning">

Note that the tokenizer used for GPT models does not need any of these tokens mentioned
above but only uses an <|endoftext|> token for simplicity

</div>

<div class="alert alert-block alert-warning">

the tokenizer used for GPT models also doesn't use an <|unk|> token for outof-vocabulary words. Instead, GPT models use a byte pair encoding tokenizer, which breaks
down words into subword units
</div>

### BYTE PAIR ENCODING (BPE)


<div class="alert alert-block alert-success">

We implemented a simple tokenization scheme in the previous sections for illustration
purposes.

This section covers a more sophisticated tokenization scheme based on a concept
called byte pair encoding (BPE).

The BPE tokenizer covered in this section was used to train
LLMs such as GPT-2, GPT-3, and the original model used in ChatGPT.</div>

<div class="alert alert-block alert-warning">

Since implementing BPE can be relatively complicated, we will use an existing Python
open-source library called tiktoken (https://github.com/openai/tiktoken).

This library implements
the BPE algorithm very efficiently based on source code in Rust.
</div>

In [ ]:
! pip3 install tiktoken

In [ ]:
import importlib
import tiktoken

print("tiktoken version:", importlib.metadata.version("tiktoken"))

tiktoken version: 0.8.0


<div class="alert alert-block alert-success">
Once installed, we can instantiate the BPE tokenizer from tiktoken as follows:</div>


In [ ]:
tokenizer = tiktoken.get_encoding("gpt2")

<div class="alert alert-block alert-success">
The usage of this tokenizer is similar to SimpleTokenizerV2 we implemented previously via
an encode method:</div>



In [ ]:
text = (
    "Hello, do you like tea? <|endoftext|> In the sunlit terraces"
     "of someunknownPlace."
)

integers = tokenizer.encode(text, allowed_special={"<|endoftext|>"})

print(integers)

[15496, 11, 466, 345, 588, 8887, 30, 220, 50256, 554, 262, 4252, 18250, 8812, 2114, 1659, 617, 34680, 27271, 13]


<div class="alert alert-block alert-info">
    
The code above prints the following token IDs:

</div>

<div class="alert alert-block alert-success">
We can then convert the token IDs back into text using the decode method, similar to our
SimpleTokenizerV2 earlier:</div>


In [ ]:
strings = tokenizer.decode(integers)

print(strings)

Hello, do you like tea? <|endoftext|> In the sunlit terracesof someunknownPlace.


<div class="alert alert-block alert-warning">

We can make two noteworthy observations based on the token IDs and decoded text
above.

First, the <|endoftext|> token is assigned a relatively large token ID, namely,
50256.

In fact, the BPE tokenizer, which was used to train models such as GPT-2, GPT-3,
and the original model used in ChatGPT, has a total vocabulary size of 50,257, with
<|endoftext|> being assigned the largest token ID.
    


</div>

<div class="alert alert-block alert-warning">

Second, the BPE tokenizer above encodes and decodes unknown words, such as
"someunknownPlace" correctly.

The BPE tokenizer can handle any unknown word. How does
it achieve this without using <|unk|> tokens?
    


</div>

<div class="alert alert-block alert-warning">

The algorithm underlying BPE breaks down words that aren't in its predefined vocabulary
into smaller subword units or even individual characters.

The enables it to handle out-ofvocabulary words.

So, thanks to the BPE algorithm, if the tokenizer encounters an
unfamiliar word during tokenization, it can represent it as a sequence of subword tokens or
characters
    


</div>

**Let us take another simple example to illustrate how the BPE tokenizer deals with unknown tokens**

In [ ]:
integers = tokenizer.encode("Akwirw ier")
print(integers)

strings = tokenizer.decode(integers)
print(strings)

[33901, 86, 343, 86, 220, 959]
Akwirw ier


### CREATING INPUT-TARGET PAIRS

<div class="alert alert-block alert-success">
In this section we implement a data loader that fetches the input-target pairs using a sliding window approach.</div>

<div class="alert alert-block alert-success">
To get started, we will first tokenize the whole The CR7 short story we worked with
earlier using the BPE tokenizer introduced in the previous section:</div>



In [ ]:
with open("cr7.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()

enc_text = tokenizer.encode(raw_text)
print(len(enc_text))

949


<div class="alert alert-block alert-info">
    
Executing the code above will return 5145, the total number of tokens in the training set,
after applying the BPE tokenizer.

</div>

<div class="alert alert-block alert-success">
Next, we remove the first 50 tokens from the dataset for demonstration purposes as it
results in a slightly more interesting text passage in the next steps:</div>

In [ ]:
enc_sample = enc_text[50:]


<div class="alert alert-block alert-success">
One of the easiest and most intuitive ways to create the input-target pairs for the nextword prediction task is to create two variables, x and y, where x contains the input tokens
and y contains the targets, which are the inputs shifted by 1:</div>

<div class="alert alert-block alert-info">
The context size determines how many tokens are included in the input

</div>



In [ ]:
context_size = 4 #length of the input
#The context_size of 4 means that the model is trained to look at a sequence of 4 words (or tokens)
#to predict the next word in the sequence.
#The input x is the first 4 tokens [1, 2, 3, 4], and the target y is the next 4 tokens [2, 3, 4, 5]

x = enc_sample[:context_size]
y = enc_sample[1:context_size+1]

print(f"x: {x}")
print(f"y:      {y}")

x: [262, 6000, 1938, 286]
y:      [6000, 1938, 286, 477]


<div class="alert alert-block alert-success">
Processing the inputs along with the targets, which are the inputs shifted by one position,
we can then create the next-word prediction tasks as
follows:</div>

In [ ]:
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]

    print(context, "---->", desired)

[262] ----> 6000
[262, 6000] ----> 1938
[262, 6000, 1938] ----> 286
[262, 6000, 1938, 286] ----> 477


<div class="alert alert-block alert-info">
Everything left of the arrow (---->) refers to the input an LLM would receive, and the token
ID on the right side of the arrow represents the target token ID that the LLM is supposed to
predict.
</div>

<div class="alert alert-block alert-success">
For illustration purposes, let's repeat the previous code but convert the token IDs into
text:</div>

In [ ]:
for i in range(1, context_size+1):
    context = enc_sample[:i]
    desired = enc_sample[i]

    print(tokenizer.decode(context), "---->", tokenizer.decode([desired]))

 the ---->  greatest
 the greatest ---->  players
 the greatest players ---->  of
 the greatest players of ---->  all


<div class="alert alert-block alert-warning">

We've now created the input-target pairs that we can turn into use for the LLM training in
upcoming chapters.
    
</div>

<div class="alert alert-block alert-warning">

There's only one more task before we can turn the tokens into embeddings:implementing an efficient data loader that
iterates over the input dataset and returns the inputs and targets as PyTorch tensors, which
can be thought of as multidimensional arrays.
    
</div>

<div class="alert alert-block alert-warning">

In particular, we are interested in returning two tensors: an input tensor containing the
text that the LLM sees and a target tensor that includes the targets for the LLM to predict,
    
</div>

### IMPLEMENTING A DATA LOADER

<div class="alert alert-block alert-success">
For the efficient data loader implementation, we will use PyTorch's built-in Dataset and
DataLoader classes.</div>

<div class="alert alert-block alert-info">
    
Step 1: Tokenize the entire text
    
Step 2: Use a sliding window to chunk the book into overlapping sequences of max_length

Step 3: Return the total number of rows in the dataset

Step 4: Return a single row from the dataset
</div>

In [ ]:
from torch.utils.data import Dataset, DataLoader


class GPTDatasetV1(Dataset):
    def __init__(self, txt, tokenizer, max_length, stride):
        self.input_ids = []
        self.target_ids = []

        # Tokenize the entire text
        token_ids = tokenizer.encode(txt, allowed_special={"<|endoftext|>"})

        # Use a sliding window to chunk the book into overlapping sequences of max_length
        for i in range(0, len(token_ids) - max_length, stride):
            input_chunk = token_ids[i:i + max_length]
            target_chunk = token_ids[i + 1: i + max_length + 1]
            self.input_ids.append(torch.tensor(input_chunk))
            self.target_ids.append(torch.tensor(target_chunk))

    def __len__(self):
        return len(self.input_ids)

    def __getitem__(self, idx):
        return self.input_ids[idx], self.target_ids[idx]

<div class="alert alert-block alert-warning">

The GPTDatasetV1 class in listing 2.5 is based on the PyTorch Dataset class.

It defines how individual rows are fetched from the dataset.

Each row consists of a number of
token IDs (based on a max_length) assigned to an input_chunk tensor.

The target_chunk
tensor contains the corresponding targets.

I recommend reading on to see how the data
returned from this dataset looks like when we combine the dataset with a PyTorch
DataLoader -- this will bring additional intuition and clarity.
    
</div>

<div class="alert alert-block alert-success">
The following code will use the GPTDatasetV1 to load the inputs in batches via a PyTorch
DataLoader:</div>

<div class="alert alert-block alert-info">
    
Step 1: Initialize the tokenizer

Step 2: Create dataset

Step 3: drop_last=True drops the last batch if it is shorter than the specified batch_size to prevent loss spikes
during training

Step 4: The number of CPU processes to use for preprocessing
    
</div>

In [ ]:
def create_dataloader_v1(txt, batch_size=4, max_length=256,
                         stride=128, shuffle=True, drop_last=True,
                         num_workers=0):

    # Initialize the tokenizer
    tokenizer = tiktoken.get_encoding("gpt2")

    # Create dataset
    dataset = GPTDatasetV1(txt, tokenizer, max_length, stride)

    # Create dataloader
    dataloader = DataLoader(
        dataset,
        batch_size=batch_size,
        shuffle=shuffle,
        drop_last=drop_last,
        num_workers=num_workers
    )

    return dataloader

<div class="alert alert-block alert-success">
    
Let's test the dataloader with a batch size of 1 for an LLM with a context size of 4,

This will develop an intuition of how the GPTDatasetV1 class and the
create_dataloader_v1 function work together: </div>

In [ ]:
with open("cr7.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()

<div class="alert alert-block alert-info">
    
Convert dataloader into a Python iterator to fetch the next entry via Python's built-in next() function
    
</div>

In [ ]:
import torch
print("PyTorch version:", torch.__version__)
dataloader = create_dataloader_v1(
    raw_text, batch_size=1, max_length=4, stride=1, shuffle=False
)

data_iter = iter(dataloader)
first_batch = next(data_iter)
print(first_batch)

PyTorch version: 2.5.1+cu121
[tensor([[   34,  1585, 10115, 36309]]), tensor([[ 1585, 10115, 36309, 23430]])]


<div class="alert alert-block alert-warning">

The first_batch variable contains two tensors: the first tensor stores the input token IDs,
and the second tensor stores the target token IDs.

Since the max_length is set to 4, each of the two tensors contains 4 token IDs.

Note that an input size of 4 is relatively small and only chosen for illustration purposes. It is common to train LLMs with input sizes of at least
256.
    
</div>

<div class="alert alert-block alert-success">
    
To illustrate the meaning of stride=1, let's fetch another batch from this dataset: </div>

In [ ]:
second_batch = next(data_iter)
print(second_batch)

[tensor([[ 1585, 10115, 36309, 23430]]), tensor([[10115, 36309, 23430, 28458]])]


<div class="alert alert-block alert-warning">

If we compare the first with the second batch, we can see that the second batch's token
IDs are shifted by one position compared to the first batch.

For example, the second ID in
the first batch's input is 367, which is the first ID of the second batch's input.

The stride
setting dictates the number of positions the inputs shift across batches, emulating a sliding
window approach
    
</div>

<div class="alert alert-block alert-warning">

Batch sizes of 1, such as we have sampled from the data loader so far, are useful for
illustration purposes.
                                                                                 
If you have previous experience with deep learning, you may know
that small batch sizes require less memory during training but lead to more noisy model
updates.

Just like in regular deep learning, the batch size is a trade-off and hyperparameter
to experiment with when training LLMs.
    
</div>

<div class="alert alert-block alert-success">
    
Before we move on to the two final sections of this chapter that are focused on creating
the embedding vectors from the token IDs, let's have a brief look at how we can use the
data loader to sample with a batch size greater than 1: </div>

In [ ]:
dataloader = create_dataloader_v1(raw_text, batch_size=8, max_length=4, stride=4, shuffle=False)

data_iter = iter(dataloader)
inputs, targets = next(data_iter)
print("Inputs:\n", inputs)
print("\nTargets:\n", targets)

Inputs:
 tensor([[   34,  1585, 10115, 36309],
        [23430, 28458, 12761,  7058],
        [   11,  4642,   319,   642],
        [ 3945, 12863,    11,   318],
        [  257, 21813,  4708, 44185],
        [   13,   679,  5341,   355],
        [  257,  2651,   329,   290],
        [37551,  1111,   262,  7420]])

Targets:
 tensor([[ 1585, 10115, 36309, 23430],
        [28458, 12761,  7058,    11],
        [ 4642,   319,   642,  3945],
        [12863,    11,   318,   257],
        [21813,  4708, 44185,    13],
        [  679,  5341,   355,   257],
        [ 2651,   329,   290, 37551],
        [ 1111,   262,  7420,  1041]])


### CREATING TOKEN EMBEDDINGS

<div class="alert alert-block alert-success">
    
Let's illustrate how the token ID to embedding vector conversion works with a hands-on
example. Suppose we have the following four input tokens with IDs 2, 3, 5, and 1:</div>

In [ ]:
input_ids = torch.tensor([2, 3, 5, 1])


<div class="alert alert-block alert-success">
    
For the sake of simplicity and illustration purposes, suppose we have a small vocabulary of
only 6 words (instead of the 50,257 words in the BPE tokenizer vocabulary), and we want
to create embeddings of size 3 (in GPT-3, the embedding size is 12,288 dimensions):

</div>

<div class="alert alert-block alert-success">
    
Using the vocab_size and output_dim, we can instantiate an embedding layer in PyTorch,
setting the random seed to 123 for reproducibility purposes:

</div>

In [ ]:
vocab_size = 6
output_dim = 3

torch.manual_seed(123)
embedding_layer = torch.nn.Embedding(vocab_size, output_dim)

<div class="alert alert-block alert-info">
    
The print statement in the code prints the embedding layer's underlying
weight matrix:
    
</div>

In [ ]:
print(embedding_layer.weight)


Parameter containing:
tensor([[ 0.3374, -0.1778, -0.1690],
        [ 0.9178,  1.5810,  1.3010],
        [ 1.2753, -0.2010, -0.1606],
        [-0.4015,  0.9666, -1.1481],
        [-1.1589,  0.3255, -0.6315],
        [-2.8400, -0.7849, -1.4096]], requires_grad=True)


<div class="alert alert-block alert-info">
    
We can see that the weight matrix of the embedding layer contains small, random values.
These values are optimized during LLM training as part of the LLM optimization itself, as we
will see in upcoming chapters. Moreover, we can see that the weight matrix has six rows
and three columns. There is one row for each of the six possible tokens in the vocabulary.
And there is one column for each of the three embedding dimensions.
    
</div>

<div class="alert alert-block alert-success">
    
After we instantiated the embedding layer, let's now apply it to a token ID to obtain the
embedding vector:

</div>

In [ ]:
print(embedding_layer(torch.tensor([3])))


tensor([[-0.4015,  0.9666, -1.1481]], grad_fn=<EmbeddingBackward0>)


<div class="alert alert-block alert-info">
    
If we compare the embedding vector for token ID 3 to the previous embedding matrix, we
see that it is identical to the 4th row (Python starts with a zero index, so it's the row
corresponding to index 3). In other words, the embedding layer is essentially a look-up
operation that retrieves rows from the embedding layer's weight matrix via a token ID.
    
</div>

<div class="alert alert-block alert-success">
    
Previously, we have seen how to convert a single token ID into a three-dimensional
embedding vector. Let's now apply that to all four input IDs we defined earlier
(torch.tensor([2, 3, 5, 1])):

</div>

In [ ]:
print(embedding_layer(input_ids))


tensor([[ 1.2753, -0.2010, -0.1606],
        [-0.4015,  0.9666, -1.1481],
        [-2.8400, -0.7849, -1.4096],
        [ 0.9178,  1.5810,  1.3010]], grad_fn=<EmbeddingBackward0>)


<div class="alert alert-block alert-info">
    
Each row in this output matrix is obtained via a lookup operation from the embedding
weight matrix
    
</div>

### POSITIONAL EMBEDDINGS (ENCODING WORD POSITIONS)

<div class="alert alert-block alert-success">

Previously, we focused on very small embedding sizes in this chapter for illustration
purposes.

We now consider more realistic and useful embedding sizes and encode the input
tokens into a 256-dimensional vector representation.

This is smaller than what the original
GPT-3 model used (in GPT-3, the embedding size is 12,288 dimensions) but still reasonable
for experimentation.

Furthermore, we assume that the token IDs were created by the BPE
tokenizer that we implemented earlier, which has a vocabulary size of 50,257:

</div>

In [ ]:
vocab_size = 50257
output_dim = 256

token_embedding_layer = torch.nn.Embedding(vocab_size, output_dim)

<div class="alert alert-block alert-info">
    
Using the token_embedding_layer above, if we sample data from the data loader, we
embed each token in each batch into a 256-dimensional vector. If we have a batch size of 8
with four tokens each, the result will be an 8 x 4 x 256 tensor.
    
</div>

<div class="alert alert-block alert-success">

Let's instantiate the data loader ( Data sampling with a sliding window),
first:

</div>

In [ ]:
max_length = 4
dataloader = create_dataloader_v1(
    raw_text, batch_size=8, max_length=max_length,
    stride=max_length, shuffle=False
)
data_iter = iter(dataloader)
inputs, targets = next(data_iter)

In [ ]:
print("Token IDs:\n", inputs)
print("\nInputs shape:\n", inputs.shape)

Token IDs:
 tensor([[   40,   367,  2885,  1464],
        [ 1807,  3619,   402,   271],
        [10899,  2138,   257,  7026],
        [15632,   438,  2016,   257],
        [  922,  5891,  1576,   438],
        [  568,   340,   373,   645],
        [ 1049,  5975,   284,   502],
        [  284,  3285,   326,    11]])

Inputs shape:
 torch.Size([8, 4])


<div class="alert alert-block alert-info">
    
As we can see, the token ID tensor is 8x4-dimensional, meaning that the data batch
consists of 8 text samples with 4 tokens each.
    
</div>

<div class="alert alert-block alert-success">

Let's now use the embedding layer to embed these token IDs into 256-dimensional
vectors:

</div>

In [ ]:
token_embeddings = token_embedding_layer(inputs)
print(token_embeddings.shape)

torch.Size([8, 4, 256])


<div class="alert alert-block alert-info">
    
As we can tell based on the 8x4x256-dimensional tensor output, each token ID is now
embedded as a 256-dimensional vector.
    
</div>

<div class="alert alert-block alert-success">

For a GPT model's absolute embedding approach, we just need to create another
embedding layer that has the same dimension as the token_embedding_layer:

</div>

In [ ]:
context_length = max_length
pos_embedding_layer = torch.nn.Embedding(context_length, output_dim)

In [ ]:
pos_embeddings = pos_embedding_layer(torch.arange(max_length))
print(pos_embeddings.shape)

torch.Size([4, 256])


<div class="alert alert-block alert-info">
    
As shown in the preceding code example, the input to the pos_embeddings is usually a
placeholder vector torch.arange(context_length), which contains a sequence of
numbers 0, 1, ..., up to the maximum input length − 1.

The context_length is a variable
that represents the supported input size of the LLM.

Here, we choose it similar to the
maximum length of the input text.

In practice, input text can be longer than the supported
context length, in which case we have to truncate the text.
    
</div>

<div class="alert alert-block alert-info">
    
As we can see, the positional embedding tensor consists of four 256-dimensional vectors.
We can now add these directly to the token embeddings, where PyTorch will add the 4x256-
dimensional pos_embeddings tensor to each 4x256-dimensional token embedding tensor in
each of the 8 batches:
    
</div>

In [ ]:
input_embeddings = token_embeddings + pos_embeddings
print(input_embeddings.shape)

torch.Size([8, 4, 256])


<div class="alert alert-block alert-warning">

The input_embeddings we created are the embedded input
examples that can now be processed by the main LLM modules
    
</div>